File: src/http-server/cta.js
import { Logger } from "../logger.js";
import { Line, Train, TrainData } from "./train.js";

export class CTA {
    apiKey;

    name = "cta";

    constructor(apiKey) {
        this.apiKey = apiKey;
    }

    getData = async () => {
        const lines = [
            "pink",
            "red",
            "orange",
            "yellow",
            "green",
            "blue",
            "purple",
            "brown",
        ];

        // converts the lines to a comma separated string with values that reflect the CTA API naming of lines
        let requestLines = lines
            .map((line) => {
                switch (line) {
                    case "pink":
                        return "pink";
                    case "red":
                        return "red";
                    case "orange":
                        return "org";
                    case "yellow":
                        return "y";
                    case "green":
                        return "g";
                    case "blue":
                        return "blue";
                    case "purple":
                        return "p";
                    case "brown":
                        return "brn";
                    default:
                        throw `Unknown line: ${line}`;
                }
            })
            .join(",");

        // fetches data from the CTA API
        return new Promise(async (resolve, reject) => {
            try {
                const response = await fetch(
                    `http://lapi.transitchicago.com//api/1.0/ttpositions.aspx?key=${this.apiKey}&rt=${requestLines}&outputType=JSON`
                );

                let data = await response.json();

                resolve(
                    new TrainData(
                        Math.round(new Date(data.ctatt.tmst).getTime() / 1000),
                        "cta",

                        data.ctatt.route
                            // converts the train data to a list of Line objects
                            .map((route, index) => {
                                if (route.train == undefined) {
                                    Logger.warn(
                                        `No train data found for ${lines[index]} line`
                                    );

                                    return [lines[index], []];
                                } else if (
                                    route.train.constructor.name != "Array"
                                ) {
                                    return new Line(lines[index], [
                                        new Train(
                                            route.train.nextStaNm,
                                            route.train.destNm,
                                            Number(route.train.trDr),
                                            Number(route.train.heading),
                                            Number(route.train.lat),
                                            Number(route.train.lon)
                                        ),
                                    ]);
                                }

                                return new Line(
                                    lines[index],
                                    route.train.map((train) => {
                                        return new Train(
                                            train.nextStaNm,
                                            train.destNm,
                                            Number(train.trDr),
                                            Number(train.heading),
                                            Number(train.lat),
                                            Number(train.lon)
                                        );
                                    })
                                );
                            })
                    )
                );
            } catch (error) {
                reject(error);
            }
        });
    };
}

File: server/cta.go
package server

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"time"
)

type ctaData struct {
	CTATT struct {
		Timestamp string `json:"tmst"`
		ErrorCode string `json:"errCd"`
		ErrorName string `json:"errNm"`
		Route     []struct {
			Name  string `json:"@name"`
			Train []struct {
				RunNumber           string `json:"rn"`
				DestinationID       string `json:"destSt"`
				DestinationName     string `json:"destNm"`
				TrainRouteDirection string `json:"trDr"`
				NextStationID       string `json:"nextStaId"`
				NextStopID          string `json:"nextStpId"`
				NextStationName     string `json:"nextStaNm"`
				PredictionGenerated string `json:"prdt"`
				ArrivalTime         string `json:"arrT"`
				IsApproaching       string `json:"isApp"`
				IsDelayed           string `json:"isDly"`
				Latitude            string `json:"lat"`
				Longitude           string `json:"lon"`
				Heading             string `json:"heading"`
			} `json:"train"`
		} `json:"route"`
	}
}

func GetCTAData(lines []string) (*TrainData, error) {

	apiKey := "00ff09063caa46748434d5fa321d048f"

	requestLines := ""
	for _, line := range lines {
		switch line {
		case "pink":
			requestLines += "pink,"
		case "red":
			requestLines += "red,"
		case "orange":
			requestLines += "org,"
		case "yellow":
			requestLines += "y,"
		case "green":
			requestLines += "g,"
		case "blue":
			requestLines += "blue,"
		case "purple":
			requestLines += "p,"
		case "brown":
			requestLines += "brn,"
		default:
			return nil, fmt.Errorf("line %s not found", line)
		}
	}
	requestLines = requestLines[:len(requestLines)-1]

	res, err := http.Get(fmt.Sprintf("http://lapi.transitchicago.com/api/1.0/ttpositions.aspx?key=%s&rt=%s&outputType=JSON", apiKey, requestLines))

	if err != nil {
		return nil, err
	}

	var data ctaData

	err = json.NewDecoder(res.Body).Decode(&data)

	if err != nil {
		return nil, err
	}

	t, err := time.Parse(time.RFC3339, data.CTATT.Timestamp+"-05:00")

	if err != nil {
		return nil, err
	}

	ret := &TrainData{
		Timestamp: t.Unix(),
		City:      "chicago",
		Lines:     map[string][]Train{},
	}

	for i, route := range data.CTATT.Route {
		for _, train := range route.Train {
			lat, err := strconv.ParseFloat(train.Latitude, 64)
			if err != nil {
				return nil, err
			}

			lon, err := strconv.ParseFloat(train.Longitude, 64)
			if err != nil {
				return nil, err
			}

			dir, err := strconv.Atoi(train.TrainRouteDirection)

			if err != nil {
				return nil, err
			}

			heading, err := strconv.Atoi(train.Heading)

			if err != nil {
				return nil, err
			}

			if _, ok := ret.Lines[lines[i]]; !ok {
				ret.Lines[lines[i]] = []Train{}
			}

			ret.Lines[lines[i]] = append(ret.Lines[lines[i]], Train{
				LineName:    lines[i],
				NextStation: train.NextStationName,
				Latitude:    lat,
				Longitude:   lon,
				Direction:   dir,
				Heading:     heading,
			})
		}
	}

	return ret, nil
}
